datarobot_batch_scoring/writer.py

```diff
@@ -410,63 +419,147 @@ def format_result_data(self, result, batch, pred_name):
+
         try:
             while True:
+                if self.abort_flag.value or self.local_abort_flag:
+                    self._ui.debug('abort requested')
+                    break
                 try:
-                    (request, batch, pred_name) = \
-                        self.writer_queue.get_nowait()
+                    if idle_cycles > 2:
+                        self.writer_status.value = b"I"
+                    else:
+                        self.writer_status.value = b"G"
+                    msg, args = self.writer_queue.get(timeout=1)
                 except queue.Empty:
-                    (request, batch, pred_name) = self.writer_queue.get()
+                    idle_cycles += 1
+                    continue
+                idle_cycles = 0
+                self.writer_status.value = b"W"
 
-                if request == QueueMsg.ERROR:
+                if msg == WriterQueueMsg.CTX_ERROR:
                     # pred_name is a message if ERROR or WARNING
                     self._ui.debug('Writer ERROR')
-                    self.ctx.save_error(batch, error=pred_name)
+                    self.ctx.save_error(args["batch"], error=args["error"])
                     continue
-                if request == QueueMsg.WARNING:
+                elif msg == WriterQueueMsg.CTX_WARNING:
                     self._ui.debug('Writer WARNING')
-                    self.ctx.save_warning(batch, error=pred_name)
+                    self.ctx.save_warning(args["batch"], error=args["error"])
                     continue
-                if batch.id == SENTINEL.id:
+                elif msg == WriterQueueMsg.SENTINEL:
                     self._ui.debug('Writer received SENTINEL')
                     break
-                result = self.unpack_request_object(request, batch)
-                if result is False:  # unpack_request_object failed
-                    continue
-
-                (out_fields, pred) = self.format_result_data(result, batch,
-                                                             pred_name)
-
-                self.ctx.checkpoint_batch(batch, out_fields, pred)
-                self._ui.debug('Writer Queue queue length: {}'
-                               ''.format(self.writer_queue.qsize()))
+                elif msg == WriterQueueMsg.RESPONSE:
+                    processed += 1
+                    batch = args["batch"]
+                    result = self.unpack_request_object(args["request"], batch)
+                    if result is False:  # unpack_request_object failed
+                        continue
+
+                    (written_fields, comb) = self.format_result_data(result,
+                                                                     batch)
+
+                    self.ctx.checkpoint_batch(batch, written_fields, comb)
+                    written += 1
+                    rows_done += batch.rows
+                    # self._ui.debug('Writer Queue queue length: {}'
```

Comment:
gr, It's not too helpful with pointing exact line :)
I'll remove commented out things